### PR TITLE
spi_edma: add edma transfer support for NXP Kinetis DSPI driver

### PR DESCRIPTION
**Pull request by hakehuang**
add EDMA support for DSPI driver, and spi_dma_loopback test

test on FRDM_K64f
build command
west build -b frdm_k64f -- -DOVERLAY_CONFIG="overlay-mcux-dspi-dma.conf"

```
*** Booting Zephyr OS build zephyr-v2.5.0-5137-gf743940ed869  ***
Running test suite test_spi
===================================================================
START - test_spi_loopback
I: SPI test on buffers TX/RX 0x20000235/0x200011da
I: SPI test slow config
I: Start complete multiple
I: CS control inhibited (no GPIO device)
I: Passed
I: Start complete loop
I: Passed
I: Start null tx
I: Passed
I: Start half start
I: Passed
I: Start half end
I: Passed
I: Skip every 4
I: Start async call
I: Passed
I: SPI test fast config
I: Start complete multiple
I: CS control inhibited (no GPIO device)
I: Passed
I: Start complete loop
I: Passed
I: Start null tx
I: Passed
I: Start half start
I: Passed
I: Start half end
I: Passed
I: Skip every 4
I: Start async call
I: Passed
I: Start complete loop
I: CS control inhibited (no GPIO device)
I: Passed
I: Start complete loop
I: CS control inhibited (no GPIO device)
I: Passed
I: All tx/rx passed
 PASS - test_spi_loopback in 0.72 seconds
===================================================================
Test suite test_spi succeeded
===================================================================
PROJECT EXECUTION SUCCESSFUL
```

### CODE REVIEW

File: drivers/spi/Kconfig.mcux_dspi [hunk 1]
```diff
@@ -1,11 +1,43 @@
 # MCUXpresso SDK SPI
```
Comment (tbursztyka) [inline drivers/spi/Kconfig.mcux_dspi:1]: change your commit title to:

drivers: spi: Update mcux driver to support edma

File: tests/drivers/spi/spi_dma_loopback/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,9 @@
+# SPDX-License-Identifier: Apache-2.0
```
Comment (tbursztyka) [inline tests/drivers/spi/spi_dma_loopback/CMakeLists.txt:1]: couldn't the tests/drivers/spi/spi_loopback test be used already for this? ST is testing their DMA support there and it's not an issue.

**Review by tbursztyka — DISMISSED** [review]
- Remove the commit which has no content 
- Do not mix code style change with feature addition (in the mcux driver). If you really want these code style changes, consider having a separate commit for these.

Not sure it's a good idea to have an entirely new test for this. DMA support is transparent, so the existing spi_loopback test should be just fine.

File: tests/drivers/spi/spi_dma_loopback/CMakeLists.txt (hunk 2 above)
Comment (hakehuang) [reply tests/drivers/spi/spi_dma_loopback/CMakeLists.txt:1]: right, I need add an overlay_dma.conf for this, as nxp spi can work with dma and without dma

File: drivers/spi/Kconfig.mcux_dspi (hunk 1 above)
Comment (hakehuang) [reply drivers/spi/Kconfig.mcux_dspi:1]: ok

**Comment by hakehuang** [comment]
*** Booting Zephyr OS build zephyr-v2.5.0-5137-gf743940ed869  ***
Running test suite test_spi
===================================================================
START - test_spi_loopback
I: SPI test on buffers TX/RX 0x20000235/0x200011da
I: SPI test slow config
I: Start complete multiple
I: CS control inhibited (no GPIO device)
I: Passed
I: Start complete loop
I: Passed
I: Start null tx
I: Passed
I: Start half start
I: Passed
I: Start half end
I: Passed
I: Skip every 4
I: Start async call
I: Passed
I: SPI test fast config
I: Start complete multiple
I: CS control inhibited (no GPIO device)
I: Passed
I: Start complete loop
I: Passed
I: Start null tx
I: Passed
I: Start half start
I: Passed
I: Start half end
I: Passed
I: Skip every 4
I: Start async call
I: Passed
I: Start complete loop
I: CS control inhibited (no GPIO device)
I: Passed
I: Start complete loop
I: CS control inhibited (no GPIO device)
I: Passed
I: All tx/rx passed
 PASS - test_spi_loopback in 0.72 seconds
===================================================================
Test suite test_spi succeeded
===================================================================
PROJECT EXECUTION SUCCESSFUL

File: tests/drivers/spi/spi_loopback/src/spi.c [hunk 3]
```diff
@@ -430,6 +430,11 @@ static int spi_rx_every_4(const struct device *dev,
 		return 0;
 	}
 
+	if (IS_ENABLED(CONFIG_DSPI_MCUX_EDMA)) {
+		LOG_INF("Skip every 4");
```
Comment (cfriedt) [inline tests/drivers/spi/spi_loopback/src/spi.c:434]: should `LOG_INF()` be removed prior to merge?

Comment (hakehuang) [reply tests/drivers/spi/spi_loopback/src/spi.c:434]: ok

File: tests/drivers/spi/spi_loopback/testcase.yaml [hunk 4]
```diff
@@ -10,3 +10,10 @@ tests:
       fixture: spi_loopback
   driver.spi.loopback.internal:
     filter: CONFIG_SPI_LOOPBACK_MODE_LOOP
+  drivers.spi_dma.loopback:
+    tags: dma
+    harness: ztest
+    extra_args: OVERLAY_CONFIG="overlay-mcux-dspi-dma.conf"
```
Comment (cfriedt) [inline tests/drivers/spi/spi_loopback/testcase.yaml:16]: Since this is effectively a test for 1 platform, which is implied by the overlay, it might make sense to add a platform component to the name of the test.

File: tests/drivers/spi/spi_dma_loopback/CMakeLists.txt (hunk 2 above)
Comment (cfriedt) [reply tests/drivers/spi/spi_dma_loopback/CMakeLists.txt]: @hakehuang - might makes sense to add "mcux" somewhere to the name of the test, because it seems to be very specific to this platform.

Comment (hakehuang) [reply tests/drivers/spi/spi_dma_loopback/CMakeLists.txt]: @cfriedt this dedicated test has been removed, I merge it to the origin case

File: tests/drivers/spi/spi_loopback/testcase.yaml (hunk 4 above)
Comment (hakehuang) [reply tests/drivers/spi/spi_loopback/testcase.yaml:16]: right, let me correct it.

File: tests/drivers/spi/spi_dma_loopback/CMakeLists.txt (hunk 2 above)
Comment (cfriedt) [reply tests/drivers/spi/spi_dma_loopback/CMakeLists.txt]: I believe that also addresses @tbursztyka's concerns, so going to make this resolved.

File: dts/bindings/spi/nxp,kinetis-dspi.yaml [hunk 5]
```diff
@@ -41,3 +41,61 @@ properties:
...
+    nxp,rx-tx-chn-share:
+      type: boolean
+      required: false
+      description: If the edma channel shared with tx and rx
+
+    pcs:
+      type: int
+      required: false
+      description: |
+        pcs slection range form 0-4
```
Comment (MaureenHelm) [inline dts/bindings/spi/nxp,kinetis-dspi.yaml:54]: This the chip select, why do we need a new property?

File: drivers/spi/Kconfig.mcux_dspi [hunk 6]
```diff
@@ -1,11 +1,43 @@
...
+	bool "ENABLE EDMA for DSPI driver"
+	depends on HAS_MCUX && HAS_MCUX_EDMA
+	help
+	  Enable the MCUX DSPI driver.
+
+config DSPI_INIT_PRIORITY
+	int "Init priority"
+	default 70
+	help
+	  Device driver initialization priority.
```
Comment (MaureenHelm) [inline drivers/spi/Kconfig.mcux_dspi:25]: Remove this and use `SPI_INIT_PRIORITY` from `drivers/spi/Kconfig` instead

File: drivers/spi/spi_mcux_dspi.c [hunk 7]
```diff
@@ -110,11 +209,344 @@ static void spi_mcux_isr(const struct device *dev)
...
+		update_tx_dma(dev);
+		update_rx_dma(dev);
+	}
+	return 0;
+}
+
+static void dma_callback(const struct device *dma_dev, void *callback_arg,
+			 uint32_t channel, int error_code)
+{
+	const struct device *dev = device_get_binding(callback_arg);
```
Comment (MaureenHelm) [inline drivers/spi/spi_mcux_dspi.c:476]: This is a slow operation that we probably don't want to do in callback, especially if the callback is invoked in interrupt context.

File: drivers/spi/spi_mcux_dspi.c [hunk 8]
```diff
@@ -143,35 +577,44 @@ static int spi_mcux_configure(const struct device *dev,
 
 	DSPI_MasterGetDefaultConfig(&master_config);
 
+	master_config.whichPcs = config->whichPcs;
```
Comment (MaureenHelm) [inline drivers/spi/spi_mcux_dspi.c:580]: this should come from `spi_cfg->slave`

File: drivers/spi/spi_mcux_dspi.c [hunk 9]
```diff
@@ -143,35 +577,44 @@ static int spi_mcux_configure(const struct device *dev,
...
 
 	ctar_config->bitsPerFrame = word_size;
 
-	ctar_config->cpol =
-		(SPI_MODE_GET(spi_cfg->operation) & SPI_MODE_CPOL)
-		? kDSPI_ClockPolarityActiveLow
-		: kDSPI_ClockPolarityActiveHigh;
+	ctar_config->cpol = (SPI_MODE_GET(spi_cfg->operation) & SPI_MODE_CPOL) ?
+				    kDSPI_ClockPolarityActiveLow :
+				    kDSPI_ClockPolarityActiveHigh;
```
Comment (MaureenHelm) [inline drivers/spi/spi_mcux_dspi.c:609]: Please move all the whitespace changes to a separate commit

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

**Comment by cfriedt** [comment]
@hakehuang  - just a few things to fix and this should be good to merge.

File: drivers/spi/Kconfig.mcux_dspi (hunk 6 above)
Comment (hakehuang) [reply drivers/spi/Kconfig.mcux_dspi:25]: ok

File: dts/bindings/spi/nxp,kinetis-dspi.yaml (hunk 5 above)
Comment (hakehuang) [reply dts/bindings/spi/nxp,kinetis-dspi.yaml:54]: right,I mis-undsertand the config->slave, should use this.

File: drivers/spi/spi_mcux_dspi.c (hunk 7 above)
Comment (hakehuang) [reply drivers/spi/spi_mcux_dspi.c:476]: ok, revert the chagnes

File: drivers/spi/spi_mcux_dspi.c (hunk 8 above)
Comment (hakehuang) [reply drivers/spi/spi_mcux_dspi.c:580]: ok, updated

File: drivers/spi/spi_mcux_dspi.c (hunk 9 above)
Comment (hakehuang) [reply drivers/spi/spi_mcux_dspi.c:609]: ok, revert changes

**Comment by cfriedt** [comment]
@tbursztyka , @MaureenHelm - looks like comments have been addressed. Please have another look when you have some time

File: drivers/spi/spi_mcux_dspi.c [hunk 10]
```diff
@@ -299,18 +863,32 @@ static const struct spi_driver_api spi_mcux_driver_api = {
...
+			DT_INST_PROP(id, continuous_sck)),		\
+		.enableRxFifoOverWrite = UTIL_AND(			\
+			DT_INST_NODE_HAS_PROP(id, rx_fifo_overwrite),	\
+			DT_INST_PROP(id, rx_fifo_overwrite)),		\
+		.enableModifiedTimingFormat = UTIL_AND(			\
+		    DT_INST_NODE_HAS_PROP(id, modified_timing_format),	\
+		    DT_INST_PROP(id, modified_timing_format)),		\
+		.is_dma_chn_shared = UTIL_AND(				\
+		    DT_INST_NODE_HAS_PROP(id, nxp_rx_tx_chn_share),	\
+		    DT_INST_PROP(id, nxp_rx_tx_chn_share)),		\
```
Comment (MaureenHelm) [inline drivers/spi/spi_mcux_dspi.c:883]: For the boolean properties, you can just use `DT_INST_PROP`, you don't need to check for existence. For the other properties, would `DT_INST_PROP_OR` work better?

File: drivers/spi/Kconfig.mcux_dspi [hunk 11]
```diff
@@ -1,11 +1,37 @@
 # MCUXpresso SDK SPI
 
 # Copyright (c) 2016, Freescale Semiconductor, Inc.
-# Copyright (c) 2017, NXP
+# Copyright (c) 2017, 2021, NXP
```
Comment (MaureenHelm) [inline drivers/spi/Kconfig.mcux_dspi:4]: Remove the duplicate `Signed-off-by: Hake Huang <hake.huang@oss.nxp.com>` in commit `drivers: spi: Update mcux driver to support edma`

File: drivers/spi/spi_mcux_dspi.c [hunk 12]
```diff
@@ -299,18 +863,32 @@ static const struct spi_driver_api spi_mcux_driver_api = {
...
 	};								\
 	DEVICE_DT_INST_DEFINE(id,					\
 			    &spi_mcux_init,				\
 			    NULL,					\
 			    &spi_mcux_data_##id,			\
 			    &spi_mcux_config_##id,			\
 			    POST_KERNEL,				\
-			    CONFIG_KERNEL_INIT_PRIORITY_DEVICE,		\
+			    CONFIG_SPI_INIT_PRIORITY,			\
```
Comment (MaureenHelm) [inline drivers/spi/spi_mcux_dspi.c:891]: This change should be a separate commit

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

File: dts/bindings/spi/nxp,kinetis-dspi.yaml [hunk 13]
```diff
@@ -41,3 +41,55 @@ properties:
...
+    nxp,rx-tx-chn-share:
+      type: boolean
+      required: false
+      description: If the edma channel shared with tx and rx
+
+    ctar:
+      type: int
+      required: false
+      description: |
+        ctar register slection range form 0-1 for master mode, 0 for slave mode
```
[the remark this answers is not in the document]
Comment (hakehuang) [reply dts/bindings/spi/nxp,kinetis-dspi.yaml:54]: sorry, let me update those error.

File: drivers/spi/spi_mcux_dspi.c [hunk 14]
```diff
@@ -46,11 +94,61 @@ static int spi_mcux_transfer_next_packet(const struct device *dev)
...
 		/* nothing left to rx or tx, we're done! */
+		LOG_DBG("spi transceive done");
 		spi_context_cs_control(&data->ctx, false);
 		spi_context_complete(&data->ctx, 0);
 		return 0;
 	}
 
+#ifdef CONFIG_DSPI_MCUX_EDMA
+
+	if (!DEV_CFG(dev)->is_dma_chn_shared) {
```
[the remark this answers is not in the document]
Comment (hakehuang) [reply drivers/spi/spi_mcux_dspi.c:105]: right, remove duplicated usage

File: drivers/spi/Kconfig.mcux_dspi (hunk 11 above)
Comment (hakehuang) [reply drivers/spi/Kconfig.mcux_dspi:4]: ok

File: drivers/spi/spi_mcux_dspi.c (hunk 12 above)
Comment (hakehuang) [reply drivers/spi/spi_mcux_dspi.c:891]: ok

File: drivers/spi/spi_mcux_dspi.c [hunk 15]
```diff
@@ -26,15 +41,48 @@ struct spi_mcux_config {
...
+	struct dma_block_config rx_dma_block;
+	struct spi_edma_config rx_dma_config;
+	struct spi_edma_config tx_dma_config;
+	int frame_size;
+	int tx_transfer_count;
+	int rx_transfer_count;
+	uint32_t whichPcs;
+#endif
+	struct spi_buf *inner_tx_buffer;
+	struct spi_buf *inner_rx_buffer;
```
[the remark this answers is not in the document]
Comment (hakehuang) [reply drivers/spi/spi_mcux_dspi.c:69]: yes, it is for emda only. let me udpate them

File: drivers/spi/spi_mcux_dspi.c (hunk 14 above)
[the remark this answers is not in the document]
Comment (hakehuang) [reply drivers/spi/spi_mcux_dspi.c:105]: right, remove duplicate usage

File: drivers/spi/spi_mcux_dspi.c (hunk 10 above)
Comment (hakehuang) [reply drivers/spi/spi_mcux_dspi.c:883]: updated

File: drivers/spi/Kconfig.mcux_dspi (hunk 11 above)
Comment (hakehuang) [reply drivers/spi/Kconfig.mcux_dspi:4]: ok

File: drivers/spi/spi_mcux_dspi.c (hunk 10 above)
Comment (hakehuang) [reply drivers/spi/spi_mcux_dspi.c:883]: updated

File: drivers/spi/spi_mcux_dspi.c (hunk 15 above)
[the remark this answers is not in the document]
Comment (hakehuang) [reply drivers/spi/spi_mcux_dspi.c:69]: ok, move to MCUX_EDMA only, done

**Comment by cfriedt** [comment]
Looks like comments are addressed again - please revisit @MaureenHelm

**Review by dleach02 — APPROVED** [review]

File: drivers/spi/spi_mcux_dspi.c [hunk 16]
```diff
@@ -26,15 +41,48 @@ struct spi_mcux_config {
...
+	int tx_transfer_count;
+	int rx_transfer_count;
+	uint32_t whichPcs;
+	struct spi_buf *inner_tx_buffer;
+	struct spi_buf *inner_rx_buffer;
+#endif
 };
 
+#define DEV_CFG(dev) ((const struct spi_mcux_config *const)dev->config)
+#define DEV_BASE(dev) ((SPI_Type *)DEV_CFG(dev)->base)
```
Comment (MaureenHelm) [inline drivers/spi/spi_mcux_dspi.c:73]: these should be removed

File: drivers/spi/spi_mcux_dspi.c [hunk 17]
```diff
@@ -110,11 +208,346 @@ static void spi_mcux_isr(const struct device *dev)
...
+static int configure_dma(const struct device *dev)
+{
+	if (DEV_CFG(dev)->is_dma_chn_shared) {
+		LOG_DBG("shard DMA request");
+		update_tx_dma(dev);
+		update_rx_dma(dev);
+	} else {
+		update_tx_dma(dev);
+		update_rx_dma(dev);
+	}
```
Comment (MaureenHelm) [inline drivers/spi/spi_mcux_dspi.c:471]: It looks like you're using `is_dma_chn_shared` only for debug purposes, so you can pull the `update_*_dma` calls outside the conditional:
```c
	if (DEV_CFG(dev)->is_dma_chn_shared) {
		LOG_DBG("shard DMA request");
	}
	update_tx_dma(dev);
	update_rx_dma(dev);
```

File: drivers/spi/spi_mcux_dspi.c [hunk 18]
```diff
@@ -110,11 +208,346 @@ static void spi_mcux_isr(const struct device *dev)
...
+static void dma_callback(const struct device *dma_dev, void *callback_arg,
+			 uint32_t channel, int error_code)
+{
+	const struct device *dev = (struct device *)callback_arg;
+	struct spi_mcux_data *data = (struct spi_mcux_data *)dev->data;
+
+	LOG_DBG("=dma call back @channel %d=", channel);
+
+	if (error_code) {
+		LOG_DBG("error happened no callback process %d", error_code);
```
Comment (MaureenHelm) [inline drivers/spi/spi_mcux_dspi.c:484]: `LOG_ERR` instead?

File: drivers/spi/spi_mcux_dspi.c [hunk 19]
```diff
@@ -282,27 +773,112 @@ static const struct spi_driver_api spi_mcux_driver_api = {
...
-	};								\
-	static struct spi_mcux_data spi_mcux_data_##id = {		\
-		SPI_CONTEXT_INIT_LOCK(spi_mcux_data_##id, ctx),		\
-		SPI_CONTEXT_INIT_SYNC(spi_mcux_data_##id, ctx),		\
+		.pcs_sck_delay =					\
+		    DT_INST_PROP_OR(id, pcs_sck_delay, 1000),		\
+		.sck_pcs_delay =					\
+		    DT_INST_PROP_OR(id, sck_pcs_delay, 1000),		\
+		.transfer_delay =					\
+		    DT_INST_PROP_OR(id, transfer_delay, 1000),		\
```
Comment (MaureenHelm) [inline drivers/spi/spi_mcux_dspi.c:869]: Why are you changing the defaults away from zero?

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

File: drivers/spi/spi_mcux_dspi.c (hunk 19 above)
Comment (hakehuang) [reply drivers/spi/spi_mcux_dspi.c:869]: this is copied from getDefault setting of SDK.

File: drivers/spi/spi_mcux_dspi.c (hunk 18 above)
Comment (hakehuang) [reply drivers/spi/spi_mcux_dspi.c:484]: ok

File: drivers/spi/spi_mcux_dspi.c (hunk 17 above)
Comment (hakehuang) [reply drivers/spi/spi_mcux_dspi.c:471]: updated

File: drivers/spi/spi_mcux_dspi.c (hunk 16 above)
Comment (hakehuang) [reply drivers/spi/spi_mcux_dspi.c:73]: ok, removed.

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

**Comment by cfriedt** [comment]
@MaureenHelm - could you please have another look?

File: drivers/spi/Kconfig.mcux_dspi [hunk 20]
```diff
@@ -1,11 +1,37 @@
...
+
+if DSPI_MCUX_EDMA
+
+config MCUX_DSPI_BUFFER_SIZE
+	int "buffer size"
+	default 144
+	help
+	  data buffer size
+
+config MCUX_DSPI_EDMA_SHUFFLE_DATA
```
Comment (mmahadevan108) [inline drivers/spi/Kconfig.mcux_dspi:29]: This config does not seem to be used, can you please check if this is needed

**Review by mmahadevan108 — APPROVED** [review]

**Review by MaureenHelm — APPROVED** [review]
